### DISCUSSION TITLE
Schemas do not appear in dataedit/schemas path

### DISCUSSION THREAD

**Discussion by pajot** [openego/oeplatform#890] [Q&A] [answered]
The schema page is empty, even though all the standard schemas are present:
```
oep_db=> \dn
      Liste der Schemas
     Name      | Eigentümer  
---------------+-------------
 _model_draft  | oep_db_user
 boundaries    | oep_db_user
 climate       | oep_db_user
 demand        | oep_db_user
 economy       | oep_db_user
 emission      | oep_db_user
 environment   | oep_db_user
 grid          | oep_db_user
 model_draft   | oep_db_user
 openstreetmap | oep_db_user
 public        | postgres
 reference     | oep_db_user
 scenario      | oep_db_user
 society       | oep_db_user
 supply        | oep_db_user
 topology      | postgres
(16 Zeilen)
```
What do I need to configure to make this work?

**Comment by wingechr** [comment, marked as answer]
I think they should appear once they are populated with tables. tables are always created first in the model draft.

@MGlauer is this correct?

**Reply by pajot** [reply]
I have a test table in model_draft. It is visible within postgresql:

```
                                                         Tabelle »model_draft.test«
  Spalte  |   Typ   | Sortierfolge | NULL erlaubt? |                 Vorgabewert                  | Speicherung | Statistikziel | Beschreibung 
----------+---------+--------------+---------------+----------------------------------------------+-------------+---------------+--------------
 id       | bigint  |              | not null      | nextval('model_draft.test_id_seq'::regclass) | plain       |               | 
 lunch    | text    |              | not null      |                                              | extended    |               | 
 quantity | integer |              | not null      |                                              | plain       |               | 
Indexe:
    "test_pkey" PRIMARY KEY, btree (id)
Zugriffsmethode: heap

```

**Reply by MGlauer** [reply]
Yes. But as @wingechr pointed out, those tables have to be registered within the OEP. This is done automatically, if the table has been created via the API. Otherwise, you will need to register them manually in `dataedit.models.Table` and `dataedit.models.Schema`. There is also the  [`mirror`](https://github.com/OpenEnergyPlatform/oeplatform/blob/develop/dataedit/management/commands/mirror.py) command (IIRC usage: `python manage.py mirror`), that crawls through the database and adds all available tables that are within whitelisted schemas.

**Reply by MGlauer** [reply]
One important note: `mirror` does also *unregister* all tables that cannot be found in the database any longer - in case this is relevant

**Reply by pajot** [reply]
@MGlauer Do tables created using the wizard require manual registration?

**Reply by wingechr** [reply]
the wizard uses the api, so you must be logged in as a registered user, but I don't think it matters how the user was created

**Reply by pajot** [reply]
Sorry, I meant registration in the sense of registration of tables in OEP. But we've cleared it up; the table registration failed initially because the schema _model_draft was absent.

**Reply by pajot** [reply]
> Yes. But as @wingechr pointed out, those tables have to be registered within the OEP. This is done automatically, if the table has been created via the API. Otherwise, you will need to register them manually in `dataedit.models.Table` and `dataedit.models.Schema`. There is also the [`mirror`](https://github.com/OpenEnergyPlatform/oeplatform/blob/develop/dataedit/management/commands/mirror.py) command (IIRC usage: `python manage.py mirror`), that crawls through the database and adds all available tables that are within whitelisted schemas.

@MGlauer I am having trouble locating `dataedit.models.Table` in PostgreSQL. Is `dataedit` a schema?

I'm trying to see how OEP registered my test table, which at least appears in the model_draft schema, even if I can't view the table structure due to the ArgumentError described in #893 .

Here's what I see:

```
oep_db=# \dn
       List of schemas
     Name      |    Owner    
---------------+-------------
 _model_draft  | oep_db_user
 boundaries    | oep_db_user
 climate       | oep_db_user
 demand        | oep_db_user
 economy       | oep_db_user
 emission      | oep_db_user
 environment   | oep_db_user
 grid          | oep_db_user
 model_draft   | oep_db_user
 openstreetmap | oep_db_user
 public        | postgres
 reference     | oep_db_user
 scenario      | oep_db_user
 society       | oep_db_user
 supply        | oep_db_user
 topology      | postgres
(16 rows)
```

```
oep_db=# \dt
                List of relations
  Schema  |      Name       | Type  |    Owner    
----------+-----------------+-------+-------------
 public   | _edit_base      | table | oep_db_user
 public   | _insert_base    | table | oep_db_user
 public   | alembic_version | table | oep_db_user
 public   | api_columns     | table | oep_db_user
 public   | api_constraints | table | oep_db_user
 public   | meta_search     | table | oep_db_user
 public   | spatial_ref_sys | table | postgres
 public   | table_tags      | table | oep_db_user
 public   | tags            | table | oep_db_user
 topology | layer           | table | postgres
 topology | topology        | table | postgres
(11 rows)
```

```
oep_django=# \dn
  List of schemas
  Name  |  Owner   
--------+----------
 public | postgres
(1 row)
```

```
oep_django=# \dt
                      List of relations
 Schema |           Name            | Type  |      Owner      
--------+---------------------------+-------+-----------------
 public | api_uploadedimages        | table | oep_django_user
 public | auth_group                | table | oep_django_user
 public | auth_group_permissions    | table | oep_django_user
 public | auth_permission           | table | oep_django_user
 public | authtoken_token           | table | oep_django_user
 public | axes_accessattempt        | table | oep_django_user
 public | axes_accesslog            | table | oep_django_user
 public | captcha_captchastore      | table | oep_django_user
 public | dataedit_filter           | table | oep_django_user
 public | dataedit_schema           | table | oep_django_user
 public | dataedit_schema_tags      | table | oep_django_user
 public | dataedit_table            | table | oep_django_user
 public | dataedit_table_tags       | table | oep_django_user
 public | dataedit_tablerevision    | table | oep_django_user
 public | dataedit_tag              | table | oep_django_user
 public | dataedit_view             | table | oep_django_user
 public | django_admin_log          | table | oep_django_user
 public | django_content_type       | table | oep_django_user
 public | django_migrations         | table | oep_django_user
 public | django_session            | table | oep_django_user
 public | login_activationtoken     | table | oep_django_user
 public | login_groupmembership     | table | oep_django_user
 public | login_grouppermission     | table | oep_django_user
 public | login_myuser              | table | oep_django_user
 public | login_usergroup           | table | oep_django_user
 public | login_userpermission      | table | oep_django_user
 public | modelview_basicfactsheet  | table | oep_django_user
 public | modelview_energyframework | table | oep_django_user
 public | modelview_energymodel     | table | oep_django_user
 public | modelview_energyscenario  | table | oep_django_user
 public | modelview_energystudy     | table | oep_django_user
 public | tutorials_tutorial        | table | oep_django_user
(32 rows)
```

**Reply by MGlauer** [reply]
`dataedit.models.Table` is a reference to the respective [django model](https://github.com/OpenEnergyPlatform/oeplatform/blob/13b984c0f02890b68556c0d708831f73bcfed554/dataedit/models.py#L51). This table should hold a register of all tables. IIRC this corresponds to the postgres table `public.dataedit_table`.

**Reply by pajot** [reply]
Okay, that makes sense now:
```
oep_django=# SELECT * FROM dataedit_table;
 id | name | schema_id | search 
----+------+-----------+--------
  1 | test |         1 | 
(1 row)

oep_django=# SELECT * FROM dataedit_schema;
 id |    name     
----+-------------
  1 | model_draft
(1 row)
```
So if I manually import data, then I need only to create an entry in `public.dataedit_schema` (if the schema does not yet exist) and `public.dataedit_table` (both in the oep_django database) to have it appear in the OEP database browser. That table itself is in `oep_db`. Have I understood this correctly?

**Reply by MGlauer** [reply]
Yes, but the default approach is to add tables via wizard or API. If those did not function properly, it is a bug. Also, the OEP has a fixed set of schemas:

    * boundaries
    * climate
    * demand
    * economy
    * emission
    * environment
    * grid
    * model_draft
    * openstreetmap
    * policy
    * reference
    * scenario
    * society
    * supply

The schema handling is currently under development and therefore subject to change, if my information is still up-to-date.

**Reply by pajot** [reply]
We have users with enormous tables that have up to 400 fields; expecting them to manually create the receiving table before importing data isn't realistic. Does the API allow automatic table creation on import based on the structure of the imported data?

**Reply by MGlauer** [reply]
Indeed, that is what the API is for. A manual creation of tables should not be necessary. You may also want to check out the [`oedialect`](https://github.com/OpenEnergyPlatform/oedialect), that has been developed to simplify the automated table creation and other API tasks.

**Reply by MGlauer** [reply]
Here is a tutorial on how to create tables via the API: https://openenergy-platform.org/tutorials/jupyter/OEP_API_tutorial_part1/

**Reply by pajot** [reply]
The tutorial refers to `api/token_config.py`:
```
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Input In [3], in <module>
      2 import pandas as pd
      3 from IPython.core.display import HTML
----> 5 from token_config import oep_url, get_oep_token
      7 # token
      8 your_token = get_oep_token()

ModuleNotFoundError: No module named 'token_config'
```
I cannot find this module anywhere. Is the token just a string?

**Comment by wingechr** [comment]
was the table created via the api?

**Reply by pajot** [reply]
I created it using the create table wizard. Does that use the API?

**Reply by wingechr** [reply]
yes

**Reply by pajot** [reply]
Okay, then there's a problem.

The table seems to have been created in model_draft when the _model_draft schema did not yet exist. Does the code rely on _model_draft to perform registration?

**Reply by wingechr** [reply]
a yes, that seems to be a plausible cause. so delete the table in the postgres backend and create it again with the wizard/api
